Add rhsm_repository module

giovannisciortino opened the pull request:
##### SUMMARY

Fixes #28159

A new module to manage subscription-manager repositories based on the discussion reported in #28159 based on the comment https://github.com/ansible/ansible/issues/28159#issuecomment-322184760 https://github.com/ansible/ansible/issues/28159#issuecomment-322184773 a new module called "redhat_repository" to go alongside "redhat_subscription" has been created. 

It allows to list enabled/disable/all repositories
It allows to execute the needs described in https://github.com/ansible/ansible/issues/28159#issuecomment-322184754
It allows to enable and disable repositories(also using wildcard * and ?) in a single task in order to perform the following tasks in a idempotent way(reporting a valid changed value in the result):
- Enable repositories foo and bar and disable all the other available repositories
- Enable all the available repositories excluding only foo and bar

It supports also the check mode

I tested the module on RHEL 6.7 ( subscription-manager-1.14.10-1.el6.x86_64 ) and RHEL 7.2 ( subscription-manager-1.15.9-15.el7.x86_64 ) with a rhsm account with about 500 repositories avaialable.
I tested the module also on RHEL 7.2 with subscription-manager register to Red Hat Satellite 6.2

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
redhat_repository

##### ANSIBLE VERSION
```
$ build/scripts-2.7/ansible --version
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]

```


##### ADDITIONAL INFORMATION
I report below some examples and their output:

**List all repositories**
```
$ ./build/scripts-2.7/ansible -b gs-client01 -m redhat_repository -a list=enabled 
gs-client01 | SUCCESS => {
    "changed": false, 
    "failed": false, 
    "repositories": [
        {
            "enabled": true, 
            "id": "rhel-7-server-optional-rpms", 
            "name": "Red Hat Enterprise Linux 7 Server - Optional (RPMs)", 
            "url": "https://sat62.localdomain/pulp/repos/Org1/Library/RHEL7/content/dist/rhel/server/7/$releasever/$basearch/optional/os"
        }, 
        {
            "enabled": true, 
            "id": "rhel-7-server-satellite-tools-6.2-rpms", 
            "name": "Red Hat Satellite Tools 6.2 (for RHEL 7 Server) (RPMs)", 
            "url": "https://sat62.localdomain/pulp/repos/Org1/Library/RHEL7/content/dist/rhel/server/7/7Server/$basearch/sat-tools/6.2/os"
        }, 
        {
            "enabled": true, 
            "id": "rhel-7-server-rpms", 
            "name": "Red Hat Enterprise Linux 7 Server (RPMs)", 
            "url": "https://sat62.localdomain/pulp/repos/Org1/Library/RHEL7/content/dist/rhel/server/7/$releasever/$basearch/os"
        }
    ]
}

```


**Enable rhel-?-server-rpms and disable all the other repositories**
```
$ ./build/scripts-2.7/ansible -b gs-client01 -m redhat_repository -a "enable=rhel-?-server-rpms disable=*"
gs-client01 | SUCCESS => {
    "changed": false, 
    "failed": false, 
    "repositories": [
        [
            {
                "enabled": true, 
                "id": "rhel-7-server-rpms", 
                "name": "Red Hat Enterprise Linux 7 Server (RPMs)", 
                "url": "https://sat62.localdomain/pulp/repos/Org1/Library/RHEL7/content/dist/rhel/server/7/$releasever/$basearch/os"
            }
        ], 
        [
            {
                "enabled": false, 
                "id": "rhel-7-server-optional-rpms", 
                "name": "Red Hat Enterprise Linux 7 Server - Optional (RPMs)", 
                "url": "https://sat62.localdomain/pulp/repos/Org1/Library/RHEL7/content/dist/rhel/server/7/$releasever/$basearch/optional/os"
            }, 
            {
                "enabled": false, 
                "id": "rhel-7-server-satellite-tools-6.2-rpms", 
                "name": "Red Hat Satellite Tools 6.2 (for RHEL 7 Server) (RPMs)", 
                "url": "https://sat62.localdomain/pulp/repos/Org1/Library/RHEL7/content/dist/rhel/server/7/7Server/$basearch/sat-tools/6.2/os"
            }
        ]
    ]
}
```

giovannisciortino commented:
I have fixed all the error reported in the failed test except the following:

> lib/ansible/modules/packaging/os/redhat_repository.py:0:0: E316 ANSIBLE_METADATA.metadata_version: not a valid value for dictionary value @ data['metadata_version']. Got '1.0'

The ANSIBLE_METADATA dictionary is equal to other modules as for instance redhat_subscription.
Please suggest me how to fix this failed test.

albertomurillo commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,275 @@
+
+- name: Enable a RHSM repository
+  redhat_subscription:
+    enable: rhel-7-server-rpms
```
This syntax is imperative telling 'what to do' vs the rest of the modules where a resource name and its desired state is given.

Why not
```
- name: Enable a RHSM repository
  redhat_repository:
    name: rhel-7-server-rpms
    state: present
```

giovannisciortino replied:
@albertomurillo thanks for your review
In the last commit I have added the list parameter as you suggest, now one(and only one) of the option list or state is required.
I have added in the documentation an example related to 'disable all task except one' based on your suggestion.

Note: For the option state I have used the values enabled/disabled instead of present/absent because "subscription-manager repos" enable/disable repositorities but it doesn't add/remove them from /etc/yum.repos.d/redhat.repo

I report below some examples with the new syntax:
```
- name: List all RHSM repositories.
  redhat_repository:
    list: all
  register: rhsm_repository

- name: List enabled RHSM repositories.
  redhat_repository:
    list: enabled
  register: enabled_rhsm_repository

- name: Enable a RHSM repository
  redhat_repository:
    name: rhel-7-server-rpms
    state: enabled

- name: Disable all RHSM repositories
  redhat_repository:
    name: '*'
    state: disabled

- name: Enable all repository starting with rhel-6-server
  redhat_repository:
    name: 'rhel-6-server*'
    state: enabled

- name: Disable all repositories except rhel-7-server-rpms
  redhat_repository:
    name: "{{ item }}"
    state: disabled
  with_items: "{{ rhsm_repository.repositories | map(attribute='id') | difference(['rhel-7-server-rpms']) }}"

```

sysadmind commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,275 @@
+    - In order to manage rhsm repositories the system must be already registered to rhsm manually or using the ansible module redhat_subscription.
+requirements:
+    - subscription-manager
+options:
```
I think it would be much more Ansible-like if there was instead a `name` option that could take multiple names and then a `state` option that could be set to `list`, `enabled`, or `disabled`. That would also eliminate the need for `disable_first` because you would instead run this module twice to get the desired effect. When set to `list`, name could be a not required parameter.

giovannisciortino replied:
I have chosen to use the current options (enable, disable, list and disable_first) to allow users to enable a set of repositories and disable all the others using a single task (I think this is a common need and a valid single subscription-manager command). If a use a single task the result variable changed will report correctly the value False(also in check mode) when I will execute the task on a host where this task has been already applied.

Instead using the Ansible-like options, I have difficult to disable and enable repositories using a single task but I have to split them in two tasks as in the following example and they will return always the result changed=True when applied multiple times on the same host.

```
- redhat_repository: name=* state=disabled 
- redhat_repository: name=rhel-6-server-rpms state= enabled
```
What do you think about this? Is it a good reason to use the current options or do you prefer Ansible-like parameters?

sysadmind replied:
I still think that the Ansible style parameters are better for the majority of use cases, although I totally get that disabling all except 1 would be hard. Ansible typically tries to be very explicit and a wildcard is not a very common option for modules for that reason. 

I'd still like to see this change, but I'd really like to hear what some of the other module owners have to say on this as I could certainly be the minority in this position.

albertomurillo replied:
I think too that ansible style parameters should be consistent with the rest of the modules. For disabling all but one repository a playbook like this can be used:

```yaml
- name: List installed repositories
  redhat_repository: list=enabled register=repos

- name: Install rhel-6-server-rpms repo
  redhat_repository: name=rhel-6-server-rpms state=present enabled=yes

- name: Remove all repositories except rhel-6-server-rpms
  redhat_repository:
    name: "{{ item }}"
    state: absent
  with_items: "{{ repos | difference(["rhel-6-server-rpms"]) }}"
```

jtyr commented:
I would rather see functionality like this be implemented in the `yum_repository` module as that module deals with the repo files. It just requires to change the logic a bit to walk through all repo files for any of the actions. Then enabling or disabling repos should be possible.

giovannisciortino committed:
Fixes #28159 : Added rhsm_repository module

giovannisciortino commented:
As suggested in the previous comments, I have modified the list of parameter accepted by this modules using now a "Ansible style" parameters format.
I  report below some examples:
```
- name: List all RHSM repositories.
  redhat_repository:
    state: list

- name: List enabled RHSM repositories.
  redhat_repository:
    name: enabled
    state: list

- name: Enable a RHSM repository
  redhat_repository:
    name: rhel-7-server-rpms
    state: enabled

- name: Disable all RHSM repositories
  redhat_repository:
    name: '*'
    state: disabled

- name: Enable all repository starting with rhel-6-server
  redhat_repository:
    name: 'rhel-6-server*'
    state: enabled
```

jtyr commented:
Please could you react to my previous comment?

giovannisciortino commented:
@jtyr : In the issue page there are already a couple of comment related to define a new ansible module for subscription-manager repository:
- [From @samdoran on June 13, 2016 19:55](https://github.com/ansible/ansible/issues/28159#issuecomment-322184756) 
- [From @minusdavid on July 23, 2016 14:57](https://github.com/ansible/ansible/issues/28159#issuecomment-322184773)

In my opinion, subscription-manager repositories and "traditional" yum repository managed by yum_repository have some differences and it could be difficult to manage both with yum_repository module only changing the yum_repository logic a bit to walk through all repo files for any of the actions.

A report below some differences between subscription-manager repositories and "traditional" yum repositories managed actually managed by yum_repository:
- In subscription-manager repositories options as baseurl,mirrorlist,description aren't required in yum_repository module and they aren't usually changed from default. In yum_repository this option are required https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/yum_repository.py#L708
- Enabling a repository using the command "subcription-manager repos --enable repo_name" execute some API calls to cdn.redhat.com, create/update ssl yum repo certificates in /etc/pki/entitlement/ . It doens't change only the parameter enabled in /etc/yum.repos.d/redhat.repo configuration file

I think the subscription-manager repository could be integrated in yum_repository module but another  options like a boolean named rhsm_repository should be addded to identify this type of repositories and manage them in a different way using "subscription-manager repos ..." commands but I think that the difficult of developing and testing this module will increase in order to manage two different type of objects.

jtyr commented:
Thanks for the explanation. If the `subcription-manager` actually does some calls back to CDN, then I think it's good to have an extra module for that.

sysadmind commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,240 @@
+          - If state is equal to 'list', retries the list of repositories and store it in 'rhsm_repositories' return value
+        choices: [ "list", "enabled", "disabled" ]
+        required: True
+        default: None
```
This does not need a default if it is required.

sysadmind commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,240 @@
+EXAMPLES = '''
+- name: List all RHSM repositories.
+  redhat_repository:
+    state: list
```
I think it would make sense on the examples with `state: list` to add a `register: variablename` to the play as a visual example of using the list.

sysadmind commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,240 @@
+    type: list
+'''
+
+from ansible.module_utils.basic import AnsibleModule
```
I'm pretty sure all of these imports can go at the top of the file like pep8 suggests. https://www.python.org/dev/peps/pep-0008/#imports

giovannisciortino replied:
I have seen that pep8 filename produces the warning "E402 module level import not at top of file". But if I move up the import Iabove the error, the test checked by shippable fails with the following error:
`lib/ansible/modules/packaging/os/redhat_repository.py:8:0: E106 Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA.
`

sysadmind replied:
My apologies. I think I missed that this was changed in modules. Thanks for the updates.

sysadmind commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,240 @@
+    name = module.params['name']
+
+    if state in ['enabled', 'disabled'] and name is None:
+        err_msg = "argument name is required when state is enabled or disabled"
```
The capitalization and punctuation here don't match other messages in this module.

sysadmind commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,240 @@
+
+    if state == 'list':
+        if isinstance(name, list) and (len(name) > 1 or name[0] not in ['all', 'enabled', 'disabled']):
+            err_msg = 'when state is list, the allowed value for the option value are: all, enabled, disabled'
```
The capitalization and punctuation don't match here either.

giovannisciortino replied:
Thanks for the review.

I have applied the suggestion reported in four of your previous five comments in my last commit just pushed.
As standard to make conform capitalization and punctuation in every error message, I used yum module as example:
- Capital letter for the first characther of each message
- No dot at the end of the message

albertomurillo commented:
```yaml
- name: List enabled RHSM repositories.
  redhat_repository:
    name: enabled
    state: list
```

It bugs me a little `name: enabled`
I think the following syntax is better

```yaml
- name: List enabled RHSM repositories.
  redhat_repository:
    state: list
    enabled: yes
```

So it will list enabled when `enabled: yes`  disabled when `enabled: no` and all when `enabled` is not defined. @sysadmind what do you think?

albertomurillo commented:
or even better

```yaml
- name: List enabled RHSM repositories.
  redhat_repository:
    list: enabled
```

Where list could be enabled|disabled|all

sysadmind commented:
I'm not opposed to either way. If changing it, I like the `list` parameter idea better than the `enabled` parameter

albertomurillo commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,239 @@
+
+def run_subscription_manager(module, arguments):
+    # Execute subuscription-manager with arguments and manage common errors
+    if module.get_bin_path('subscription-manager'):
```
There is no need to call `module.get_bin_path()` twice

You could write it as follows:
```python
rhsm_bin = module.get_bin_path('subscription-manager')
if not rhsm_bin:
  module.fail_json(msg='The executable file subscription-manager was not found in PATH')
```

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+module: redhat_repository
+short_description: Manage RHSM repositories using the C(subscription-manager) command
+description:
+    - Manage(List/Enable/Disable) RHSM repositories to the Red Hat Subscription Management entitlement platform using the C(subscription-manager) command
```
Standard YAML indentation is 2 spaces. The `EXAMPLES` block is using it so please change the indentation of the `DOCUMENTATION` block to 2 spaces as well.

Also please wrap all the lines in the `DOCUMENTATION` block on 80 characters. And please put period at the end of the description.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+
+RETURN = '''
+repositories:
+    description: if state is equal to list, rhsm_repositories is the list of repositories
```
Indentation issue like in the `DOCUMENTATION` block. Missing capital letter at the beginning and period at the end of the sentence.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+        default: None
+    name:
+        description:
+          The id of repositories to enable. To operate on several repositories this can accept a comma separated list or a list.
```
`id` should be `ID`.

`a comma separated list or a list` should be `a comma separated list or a YAML list`

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+    state:
+        description:
+          state indicates the desired repositories states.
+        choices: [ "enabled", "disabled" ]
```
The values should not be quoted. And no spaces at the beginning ant the end of the list.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+description:
+    - Manage(List/Enable/Disable) RHSM repositories to the Red Hat Subscription Management entitlement platform using the C(subscription-manager) command
+version_added: "2.4"
+author: "Giovanni Sciortino (giovannibattistasciortino@gmail.com)"
```
Please use your GitHub ID instead of e-mail here.

albertomurillo replied:
or both

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+author: "Giovanni Sciortino (giovannibattistasciortino@gmail.com)"
+notes:
+    - In order to manage rhsm repositories the system must be already registered to rhsm manually or using the ansible module redhat_subscription.
+    - One option between state and list must be defined, both options in the same task must not be defined
```
Missing period.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+notes:
+    - In order to manage rhsm repositories the system must be already registered to rhsm manually or using the ansible module redhat_subscription.
+    - One option between state and list must be defined, both options in the same task must not be defined
+    - When the option state is defined, the option name must be defined
```
Missing period.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+options:
+    state:
+        description:
+          state indicates the desired repositories states.
```
Capital letter at the beginning.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+    list:
+        description:
+          List all/enabled/disabled repository.
+        choices: [ "all", "enabled", "disabled" ]
```
The values should not be quoted. And no spaces at the beginning ant the end of the list.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+import re
+import os
+from fnmatch import fnmatch
+
```
You can remove this empty line.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+    state = module.params['state']
+
+    if state is None and list is None:
+        err_msg = "Use list or state option, one of them is required"
```
What about to set `err_msg = None` just above the first `if` condition, change all following `if`s to `elif` and then remove the all `module.fail_json(msg=err_msg)` and put one extra `if` at the end to fail when `err_msg is not None`? Something like this:

```
err_msg = None

if state is None and list is None:
    err_msg = "Use list or state option, one of them is required"
elif state is not None and list is not None:
    err_msg = "Both list and state option are defined, use only one of them in the task"
elif list is not None and name is not None:
    err_msg = "When the option list is defined, the option name must not be defined"
elif state in ['enabled', 'disabled'] and name is None:
    err_msg = "When state is enabled or disabled, the option name is required"

if err_msg is not None:
    module.fail_json(msg=err_msg)
```

That should save couple of lines.

albertomurillo replied:
You can use the following options for the `AnsibleModule` which does that validation for you.

```python
        required_one_of=[["name", "list"]],
        mutually_exclusive=[["name", "list"]],
```

You can take a look at the [swupd.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/swupd.py#L299) module for reference.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+        ),
+        supports_check_mode=True
+    )
+    list = module.params['list']
```
I think you should not use `list` as a name of a variable because it is a keyword for a Python type. You can use `lst` instead.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+
+- name: Enable all repository starting with rhel-6-server
+  redhat_repository:
+    name: 'rhel-6-server*'
```
You don't have to quote the value.

albertomurillo commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,260 @@
+            name=dict(type="list",
+                      required=False),
+            state=dict(choices=['enabled', 'disabled'],
+                       default=None,
```
The default value should be enabled.

albertomurillo replied:
Also the convention is to be `present` and `absent` the difference in the `yum_repository` module is that
`present|absent` means the repo is installed and `enabled|disabled` means the repo is in use. With yum repos you can have a present repository but in a disabled state meaning the .repo file is there but there is a flag in it telling yum not to use it.

albertomurillo replied:
the `apt_repository` module also just uses `present|absent` for the `state` option

giovannisciortino replied:
I will work on "The default value should be enabled." but state value will remain enabled|disabled because as you wrote:

`present|absent means the repo is installed and enabled|disabled means the repo is in use. With yum repos you can have a present repository but in a disabled state meaning the .repo file is there but there is a flag in it telling yum not to use it.`

and the command "subscription-manager repos ...." enable/disable the rhsm yum repository, it doesn't install/remove repository entry from .repo file

albertomurillo replied:
maybe it makes sense to accept both present/absent (the standard in ansible modules) and enabled/disabled (which make sense in the subscription manager context) but would like to hear thoughts from others first.

There are some modules that accept different names for the same action where context makes more sense like [slackpg](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/slackpkg.py#L192) and [apk](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/apk.py#L304)

albertomurillo replied:
@jtyr @sysadmind ^

giovannisciortino commented:
Thanks for this review, with the last commit I have applied all the previous suggestions except:
- "In almost all modules the default is expected to be enabled." because in redhat_repository module the option list and state are mutually exclusive. If I assign a default value to list I can't enable repositories and viceversa
- https://github.com/ansible/ansible/pull/28292#discussion_r134698188 because I have replaced these error messages with built-in functions mutually_exclusive,required_one_of,required_if

albertomurillo commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+                       default=None,
+                       required=False),
+        ),
+        required_one_of=[["list", "state"]],
```
`required_one_of=[["list", "state"]],` allows the following invalid expression

```yaml
- name: Invalid task
  redhat_repository:
    state: present
```

It should be replaced with `required_one_of=[["list", "name"]]` to allow the following expresions

```yaml
- name: List repositories
  redhat_repository:
    list: all
  register: repos
```

and 

```yaml
- name: Enable myrepo
  redhat_repository:
    name: myrepo
```

Usually `state` defaults to `enable` or `present` and omitted from plays and only specified when the desired state is `absent` or `disabled`

albertomurillo commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+        ),
+        required_one_of=[["list", "state"]],
+        mutually_exclusive=[["list", "state"], ["list", "name"]],
+        required_if=[["state", "present", ["name"]]],
```
This one is not required once `required_one_of=[["list", "name"]]` and the default for `state` is `enabled`

albertomurillo commented:
@giovannisciortino I meant assigning the default value of `enabled` to the `state` option not the `list` option also `list` and `name` should be mutually exclusive not `list` and `state` Sorry for not stating it in a clearer way before.

albertomurillo commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+    default: None
+  list:
+    description:
+      - List all/enabled/disabled repository.
```
s/repository/repositories

jtyr replied:
And the `all`, `enabled` and `disabled` should be decorated as an option value.

albertomurillo commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+    name: '*'
+    state: disabled
+
+- name: Enable all repository starting with rhel-6-server
```
s/repository/repositories

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+                   command
+description:
+  - Manage(List/Enable/Disable) RHSM repositories to the Red Hat Subscription
+    Management entitlement platform using the C(subscription-manager) command
```
Missing period at the end of the sentence.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+    description:
+      - The ID of repositories to enable.
+      - To operate on several repositories this can accept a comma separated
+        list or a YAML list
```
Missing period at the end of the sentence.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+author: "Giovanni Sciortino (@giovannisciortino)"
+notes:
+  - In order to manage rhsm repositories the system must be already registered
+    to rhsm manually or using the ansible module redhat_subscription.
```
Both `rhsm` should be in uppercase letters.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+    choices: [all, enabled, disabled]
+    required: False
+    default: None
+
```
This blank line should not be here.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+    name: "{{ item }}"
+    state: disabled
+  with_items: "{{ rhsm_repository.repositories | map(attribute='id') | difference(['rhel-7-server-rpms']) }}"
+
```
This blank line should not be here.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+
+
+def get_repository_list(module, list_parameter):
+    # Generate rhsm repository list and return a list of dict
```
The `rhsm` should be in uppercase letters.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+
+
+def repository_list(module, list_parameter):
+    # Get rhsm repository list and format it for the user
```
The `rhsm` should be in uppercase letters.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+  - Manage(List/Enable/Disable) RHSM repositories to the Red Hat Subscription
+    Management entitlement platform using the C(subscription-manager) command
+version_added: "2.4"
+author: "Giovanni Sciortino (@giovannisciortino)"
```
This doesn't have to be quoted.

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,253 @@
+  - One option between state and list must be defined, both options in the
+    same task must not be defined.
+  - When the option state is defined, the option name must be defined.
+
```
Remove one blank line here.

giovannisciortino commented:
Thank for the review.
I have fixed all the previous suggestions except https://github.com/ansible/ansible/pull/28292#discussion_r134861501 (if someone want to contribute to this  discussion with his opinion is welcome )

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,247 @@
+    if not rhsm_bin:
+        module.fail_json(msg='The executable file subscription-manager was not found in PATH')
+
+    rc, out, err = module.run_command(rhsm_bin + " " + " ".join(arguments))
```
What about to use this:

```python
rc, out, err = module.run_command("%s %s" % (rhsm_bin, " ".join(arguments)))
```

giovannisciortino replied:
Modified in the last commit

albertomurillo commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,247 @@
+        ),
+        required_one_of=[["list", "name"]],
+        mutually_exclusive=[["list", "state"], ["list", "name"]],
+        required_if=[["state", "present", ["name"]]],
```
This line is no longer required since the default is already `enabled`

giovannisciortino replied:
This line has been removed in the last commit

albertomurillo commented:
@giovannisciortino The PR is almost ready. Be sure to bring it to the next [ansible meeting](https://github.com/ansible/community/tree/master/meetings) by adding a comment on the latest [public core meeting agenda](https://github.com/ansible/community/issues/220)
 and ask for comments on #ansible-devel

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,247 @@
+        if repo_enabled_re:
+            repo_enabled = repo_enabled_re.group(1)
+
+            repo = {"id": repo_id, "name": repo_name, "url": repo_url, "enabled": True if repo_enabled == '1' else False}
```
What about to write this like this:

```python
repo = {
    "id": repo_id,
    "name": repo_name,
    "url": repo_url,
    "enabled": True if repo_enabled == '1' else False
}
```

giovannisciortino replied:
Changed in the last commit 3045968

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,247 @@
+    module = AnsibleModule(
+        argument_spec=dict(
+            list=dict(choices=['all', 'enabled', 'disabled'],
+                      required=False),
```
The `required=False` is set by default so you don't have to set it explicitly. The same bellow.

giovannisciortino replied:
Changed in the last commit 3045968

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,247 @@
+        argument_spec=dict(
+            list=dict(choices=['all', 'enabled', 'disabled'],
+                      required=False),
+            name=dict(type="list",
```
Please change the double quotes for single quotes to unify the look with the rest of the code. The same bellow.

giovannisciortino replied:
Changed in the last commit 3045968

jtyr commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,247 @@
+options:
+  state:
+    description:
+      - The option state indicates the desired repositories states.
```
Please change it to `Indicates the desired repository state.`

giovannisciortino replied:
Changed in the last commit 3045968

albertomurillo commented on lib/ansible/modules/packaging/os/redhat_repository.py:
```diff
@@ -0,0 +1,248 @@
+    description:
+      - Indicates the desired repository state.
+    choices: [enabled, disabled]
+    required: False
```
required is usually defined only when true, it defaults to false so there is no need to put it explicitly

jtyr replied:
Not true. The `required` should always be present as per the [docs](http://ansible-docs.readthedocs.io/zh/stable-2.0/rst/developing_modules.html#module-checklist).

albertomurillo replied:
@jtyr You are right! `required should always be present, be it true or false`

It wasn't the case the last time I read those docs

albertomurillo replied:
So the documentation contradicts itself, which one is the most recent? [developing_modules_documenting](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html)
`required: Only needed if true, otherwise it is assumed to be false.`

jtyr replied:
Well it depends whether you talk about the `required: False` in the `DOCUMENTATION` block OR about the `required=False` in the `AnsibleModule()` object in the `main()` function. In the first case it must be always present, in the second case only if it's `True`.

bcoca replied:
the `list` option breaks how other modules deal with this a) always return the info b) _facts module c) list/query state (moving away from this).

bcoca replied:
in a) i would return the full list with the 'status' of each repo

giovannisciortino replied:
I hope to have correctly understood the suggestion of @bcoca in his last comment, I have applied the solution a) in the last commit b7412e6
 . 
Now the result variable "repositories"(containing the updated repository states) is returned always from this module when is used with any parameter(name or list).

bcoca replied:
+1 ... for brownie points add --diff support.

but noticed, you still have the 'list' option, even state=list might be better (as list shortcuts states present and absent)

giovannisciortino replied:
Added --diff support in the last commit
```
Example:
$ ./bin/ansible gs-client01 -m redhat_repository -a "state=enabled name='rhel-7-server-rpms'" -b --diff 
--- before: RHSM repositories
+++ after: RHSM repositories
@@ -1 +1 @@
-Repository 'rhel-7-server-rpms' is disabled for this system
+Repository 'rhel-7-server-rpms' is enabled for this system

gs-client01 | SUCCESS => {
    "changed": true, 
    "failed": false, 
    "repositories": [
        {
            "enabled": false, 
            "id": "rhel-7-server-optional-rpms", 
            "name": "Red Hat Enterprise Linux 7 Server - Optional (RPMs)", 
            "url": "https://gs-sat61.localdomain/pulp/repos/Org1/Library/content/dist/rhel/server/7/$releasever/$basearch/optional/os"
        }, 
        {
            "enabled": false, 
            "id": "rhel-7-server-satellite-tools-6.2-rpms", 
            "name": "Red Hat Satellite Tools 6.2 (for RHEL 7 Server) (RPMs)", 
            "url": "https://gs-sat61.localdomain/pulp/repos/Org1/Library/content/dist/rhel/server/7/7Server/$basearch/sat-tools/6.2/os"
        }, 
        {
            "enabled": true, 
            "id": "rhel-7-server-rpms", 
            "name": "Red Hat Enterprise Linux 7 Server (RPMs)", 
            "url": "https://gs-sat61.localdomain/pulp/repos/Org1/Library/content/dist/rhel/server/7/$releasever/$basearch/os"
        }
    ], 
    "results": [
        "Repository 'rhel-7-server-rpms' is enabled for this system."
    ]
}
```

Removed list option, now the option state accepts the following parameters: enabled, disabled, list, list_enabled, list_disabled. When state is equal to enabled or disabled, the option name allow to define the list of repositories to be enabled/disabled.

alikins commented:
@giovannisciortino Would like to see some test cases, but understand it is hard to mock out.

giovannisciortino commented:
@alikins the subscription-manager command(the unique command called from my ansible module) and their responses maybe could be simulated by a script providing the correct output for a fixed set of input used in the test cases. 
I can work on it if there is a concrete interest to merge it in ansible because I'm observing that proposing a new module supported by community requires more efforts than fix bugs in ansible core part. 
Can module testing be tracked as separated issue after the eventual merge of this module?

abadger commented:
Talked about this in the meeting today and noted that there is an existing PR for a module configuring redhat repos: https://github.com/ansible/ansible/pull/19717/files.  Discussion in the meeting and with the  RHSM folks is that this one is preferred for the following reasons:
* This one only operates idempotently
* there's less code in main() which is better organization which would be easier to unittest at some future date 
* already complies with the singular name rule

@jfenal, if you and @giovannisciortino could collaborate, though, and see if there's other things that you thought of in your PR that could be added here, that would be appreciated.  (Noting that, as stated above, idempotent-only was seen as a benefit of this PR).

@alikins can do final sanity check and merge when you've had a chance to look at each other's code for ideas to merge in here.

dmsimard commented:
I don't want to bikeshed too much but Is ``redhat_repository`` the right name for this ?

Like, pretend there would be a ``yum_repository`` module, a ``dnf_repository`` module or even an ``apt_repository`` module (_those would be awesome btw_)... or even a plain ``repository`` module that'd wrap around those just like ``package`` wraps around ``yum``, ``dnf`` or ``apt``.

Would this module name still be appropriate ? It sounds like this should really be ``rhsm_repository``.

dmsimard commented:
It turns out that yum_repository and apt_repository are a thing, actually. So, there's that.

dagwieers commented:
@dmsimard I also object to the name, mostly because if Red Hat decides to use a complete different way of managing repositories or packages, we're stuck with this name. It should be called after the technology being used rather than a vendor-name. So probably **rhsm_repository** ?

PS Yes, Red Hat has switched from yum to dnf,  and from RHN to Subscription Manager, so this is not a theoretical example :-)

giovannisciortino commented:
Based on the previous two comments I renamed,in the last commit, the module and this PR from redhat_repository to rhsm_repository. I'm agree with them, the technology to manage RH repository could change in the future.
The unique drawback of this solution is the existence of the module redhat_subscription (already present in ansible) to manage the subscription using subscription-manager but this isn't related to this module.

Furthermore as requested in this [comment](https://github.com/ansible/ansible/pull/28292#issuecomment-331275019) I analyzed the PR https://github.com/ansible/ansible/pull/19717/files to find possible new feature to include in my PR but I haven't found any other feature not present already in my PR but I'm open to any suggestion.

jtyr commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,260 @@
+        if repo_enabled_re:
+            repo_enabled = repo_enabled_re.group(1)
+
+            repo = {"id": repo_id,
```
Please format this like this:

```python
            repo = {
                "id": repo_id,
                "name": repo_name,
                "url": repo_url,
                "enabled": True if repo_enabled == '1' else False
            }
```

giovannisciortino replied:
Changed in the last commit, thanks for the review

jtyr commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,260 @@
+  rhsm_repository:
+    name: "{{ item }}"
+    state: disabled
+  with_items: "{{ rhsm_repository.repositories | map(attribute='id') | difference(['rhel-7-server-rpms']) }}"
```
What about to format this like this:

```yaml
  with_items: "{{
    rhsm_repository.repositories |
    map(attribute='id') |
    difference(['rhel-7-server-rpms']) }}"
```

giovannisciortino replied:
Changed in the last commit, thanks for the review

jtyr commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,260 @@
+            'before_header': "RHSM repositories",
+            'after_header': "RHSM repositories"}
+
+    if module.check_mode:
```
What about to simplify this like this:

```python
    if not module.check_mode:
        rc, out, err = run_subscription_manager(module, rhsm_arguments)
        results = out.split('\n')

    module.exit_json(results=results, changed=changed, repositories=updated_repo_list, diff=diff)
```

giovannisciortino replied:
Changed in the last commit, thanks for the review

jtyr commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,260 @@
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(type='list'),
+            state=dict(choices=['enabled', 'disabled', 'list', 'list_enabled', 'list_disabled'],
```
Please format this like this:

```python
            state=dict(
                choices=['enabled', 'disabled', 'list', 'list_enabled', 'list_disabled'],
                default='enabled'),
```

giovannisciortino replied:
Changed in the last commit, thanks for the review

jtyr commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,263 @@
+        rhsm_arguments = ['repos', '--list']
+    rc, out, err = run_subscription_manager(module, rhsm_arguments)
+
+    skip_lines = ['+----------------------------------------------------------+',
```
Please format this like this:

```python
    skip_lines = [
        '+----------------------------------------------------------+',
        '    Available Repositories in /etc/yum.repos.d/redhat.repo'
    ]
```

giovannisciortino replied:
Changed in the last commit, thanks for the review

giovannisciortino commented:
This PR is stale from about 1 month and half, I have invested a lot of time in it and I believe I have replied to all the review modifying the code as requested. Are there some chance to have it merged or can I do some action to improve the merge probability?

sysadmind commented:
shipit

dagwieers commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,265 @@
+#!/usr/bin/python
+
+# Giovanni Sciortino (@giovannisciortino)
```
A copyright statement looks like:
```python
# Copyright: (c) 2017, Giovanni Sciortino (@giovannisciortino)
```

giovannisciortino replied:
Fixed in the last commit

dagwieers commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,265 @@
+DOCUMENTATION = '''
+---
+module: rhsm_repository
+short_description: Manage RHSM repositories using the C(subscription-manager)
```
Put this on a single line. Not sure if we want C(...) in the short_description.

giovannisciortino replied:
Fixed in the last commit

dagwieers commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,265 @@
+      - If state is equal to list, list_enabled, or list_disabled, list
+        C(all)/C(enabled)/C(disabled) repositories.
+    choices: [enabled, disabled, list, list_enabled, list_disabled]
+    required: False
```
If it is not required, you don't have to state it as it's the default.

giovannisciortino replied:
state option is required and it has the default option value "enabled". The code was already correct, I have fixed the documentation.

dagwieers commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,265 @@
+        list or a YAML list.
+      - If state is equal to enabled or disabled, this option is required.
+      - If state is not equal to enabled or disabled, this option must not be used.
+    required: False
```
Same here.

giovannisciortino replied:
Fixed, remove default value from doc

dagwieers commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,265 @@
+      - If state is equal to enabled or disabled, this option is required.
+      - If state is not equal to enabled or disabled, this option must not be used.
+    required: False
+    default: None
```
The default value is "None", so this is not needed.

giovannisciortino replied:
Fixed, remove default value from doc

dagwieers reviewed (approved):
LGTM

samdoran commented on lib/ansible/modules/packaging/os/rhsm_repository.py:
```diff
@@ -0,0 +1,263 @@
+EXAMPLES = '''
+- name: List all RHSM repositories.
+  rhsm_repository:
+    state: list
```
Listing data should be in a lookup plugin, not a module.

samdoran replied:
Or a facts plugin would also work.

samdoran committed:
Streamline module

Remove functions that should be in a facts or lookup plugin. Update docs to reflect changes in code.
